### PR TITLE
Add postfix field to HttpFileStorage

### PR DESCRIPTION
**Pull request by KorSin**

### CODE REVIEW

File: model/src/main/kotlin/config/HttpFileStorageConfiguration.kt [hunk 1]
```diff
@@ -28,6 +29,15 @@ data class HttpFileStorageConfiguration(
      */
     val url: String,
 
+    /**
+     * The postfix that is appended to the combination of the url and some additional path. Some storages process
+     * authentication via parameters that are within the final URL, so certain credentials can be stored in this
+     * postfix, e.g, "?user=standard&pwd=123". Thus, the final URL could be
+     * "https://example.com/storage/path?user=standard&pwd=123".
+     */
+    @JsonProperty("urlPostfix")
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:38]: Keep the default snake case like in other configuration classes. Instead, consider not serializing this property if it is empty.

**Review by mnonnenmacher — CHANGES_REQUESTED** [review]
Please also update the reference.conf.

File: model/src/test/kotlin/config/HttpFileStorageConfigurationTest.kt [hunk 2]
```diff
@@ -26,13 +26,18 @@ import org.ossreviewtoolkit.model.yamlMapper
...
+            "urlPostfix",
+            mapOf("key1" to "value1", "key2" to "value2")
+        )
 
         val yaml = yamlMapper.writeValueAsString(config).trim()
 
         yaml shouldBe """
             ---
             url: "url"
+            urlPostfix: "urlPostfix"
```
[the remark this answers is not in the document]
Comment (KorSin) [reply model/src/test/kotlin/config/HttpFileStorageConfigurationTest.kt:40]: This sounds reasonable. However, some things are mutual exclusive. For instance, masking the postfix and ignoring it if it is empty does not work. If it is empty, it is masked and the mask itself is empty, hence it is not ignored when serialized. One could also implement it such it is only masked when it is not empty. What is in your opinion the better solution?

[the remark this answers is not in the document]
Comment (mnonnenmacher) [reply model/src/test/kotlin/config/HttpFileStorageConfigurationTest.kt]: > One could also implement it such it is only masked when it is not empty.

That sounds like the best option to me if it makes it possible to not serialize the property if it is empty.

File: model/src/main/kotlin/config/HttpFileStorageConfiguration.kt [hunk 3]
```diff
@@ -19,15 +19,26 @@
...
 data class HttpFileStorageConfiguration(
     /**
      * The URL of the HTTP server, e.g. "https://example.com/storage".
      */
     val url: String,
 
+    /**
+     * The postfix that is appended to the combination of the url and some additional path. Some storages process
+     * authentication via parameters that are within the final URL, so certain credentials can be stored in this
+     * postfix, e.g, "?user=standard&pwd=123". Thus, the final URL could be
```
Comment (sschuberth) [inline model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:36]: Ah, so the "postfix" is just a regular query string! Shouldn't we then also simply refer to it as such? Or are there real-world cases where this indeed needs to be a generic postfix, and *not* a query string?

File: model/src/test/kotlin/config/HttpFileStorageConfigurationTest.kt [hunk 4]
```diff
@@ -38,4 +38,17 @@ class HttpFileStorageConfigurationTest : StringSpec({
               key2: "***"
             """.trimIndent()
     }
+
+    "Postfix should be masked in serialization" {
+
```
Comment (sschuberth) [inline model/src/test/kotlin/config/HttpFileStorageConfigurationTest.kt:43]: Nit: Superfluous blank line.

File: utils/src/main/kotlin/storage/HttpFileStorage.kt [hunk 5]
```diff
@@ -41,6 +41,14 @@ class HttpFileStorage(
      */
     val url: String,
 
+    /**
+     * The postfix that is appended to the combination of the url and some additional path. Some storages process
+     * authentication via parameters that are within the final URL, so certain credentials can be stored in this
+     * postfix, e.g, "?user=standard&pwd=123". Thus, the final URL could be
+     * "https://example.com/storage/path?user=standard&pwd=123".
+     */
+    val urlPostfix: String = "",
```
Comment (sschuberth) [inline utils/src/main/kotlin/storage/HttpFileStorage.kt:50]: If we don't call this `query` now anyway, I'd propose to call it just `postfix`, as `path` is also not called `urlPath`.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/config/HttpFileStorageConfiguration.kt (hunk 3 above)
Comment (KorSin) [reply model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:36]: Right now I cannot think of other cases than queries. For me it would be alright to rename it. Do you prefer `query` or `queryString` ?

Comment (sschuberth) [reply model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:36]: I have a slight preference for just `query`.

File: model/src/main/kotlin/config/HttpFileStorageConfiguration.kt [hunk 6]
```diff
@@ -37,5 +48,5 @@ data class HttpFileStorageConfiguration(
 )
 
 class MaskStringConverter : StdConverter<String, String>() {
-    override fun convert(value: String) = "***"
+    override fun convert(value: String) = if (value.isNotEmpty()) "***" else ""
```
Comment (sschuberth) [inline model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:51]: BTW, is this change really required? If so, why?

Comment (KorSin) [reply model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:51]: It was suggested by @mnonnenmacher that the field is only masked if not empty and is not printed else. Although he meant this only for the postfix I did not want to create two converters and applied this for the headers, too.

File: model/src/main/kotlin/config/HttpFileStorageConfiguration.kt (hunk 3 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:36]: And also remember to reword the commit message accordingly, and BTW, also change "Url" to "URL" there, please.

File: model/src/main/kotlin/config/HttpFileStorageConfiguration.kt (hunk 6 above)
Comment (sschuberth) [reply model/src/main/kotlin/config/HttpFileStorageConfiguration.kt:51]: I see, so this is more or less a trick to get an empty string here that you don't need to serialize. This should be explained in the commit message.

File: model/src/main/kotlin/config/FileStorageConfiguration.kt [hunk 7]
```diff
@@ -48,7 +48,11 @@ data class FileStorageConfiguration(
         }
 
         return httpFileStorage?.let { httpFileStorageConfiguration ->
-            HttpFileStorage(httpFileStorageConfiguration.url, httpFileStorageConfiguration.headers)
+            HttpFileStorage(
```
Comment (sschuberth) [inline model/src/main/kotlin/config/FileStorageConfiguration.kt:51]: Commit message:

> base Url and the query string. Furthermore, mask the query string for printing

"base URL"

> Furthermore, mask the query string for printing only if it is not empty. Otherwise it is not printed at all. Apply this to headers in the HttpFileStorageConfiguration as well.

I'd suggest instead: "Also, mask the query string only if it is not empty to be able to omit empty query strings from the serialization."

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]
